File: src/main/java/frc/robot/commands/Drivetrain_Commands/AlignShooter.java
package frc.robot.commands.Drivetrain_Commands;

import edu.wpi.first.math.MathUtil;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.DrivetrainFalcon;
import java.util.function.DoubleSupplier;

public class AlignShooter extends Command {

    private double loopIdx = 0;
    private double delta = 0.0;

    private DoubleSupplier measurementSource;
    private DoubleSupplier setpointSource;
    private DrivetrainFalcon drivetrain;

    public AlignShooter(DoubleSupplier measurementSource,
                        DoubleSupplier setpointSource,
                        DrivetrainFalcon drivetrain) {
        this.measurementSource = measurementSource;
        this.setpointSource = setpointSource;
        this.drivetrain = drivetrain;
        addRequirements(drivetrain);
    }

    @Override
    public void initialize() {
        System.out.println("INFO: AlignShooter.initialize");
        delta = MathUtil.inputModulus(setpointSource.getAsDouble() - measurementSource.getAsDouble(), -180.0, 180.0);
    }

    @Override
    public void execute() {

        delta = MathUtil.inputModulus(setpointSource.getAsDouble() - measurementSource.getAsDouble(), -180.0, 180.0);

        double output = 0.0;

        if (Math.abs(delta) > 45) {
            output = Math.signum(delta) * 0.575;
        } else if(Math.abs(delta) > 30) {
            output = Math.signum(delta) * 0.375;
        } else if(Math.abs(delta) > 1) {
            output = Math.signum(delta) * 0.26;
        }
        drivetrain.drive(0.0, -output);

//        // Debugging values
//        loopIdx++
//        if (loopIdx == 10) {
//            loopIdx = 0
//            println("INFO: Yaw: ${m_measurement.asDouble}, Heading: ${m_setpoint.asDouble}, PidOut: $pidOut, FF: $ff, TotalOut: ${pidOut+ff}")
//            println("INFO: Yaw: ${m_measurement.asDouble}, Heading: ${m_setpoint.asDouble}, Delta: $delta")
//            println("INFO: Yaw: ${m_measurement.asDouble}, Heading: ${m_setpoint.asDouble}, Output value: ${pidOut + ff}")
//
//        }
    }

    @Override
    public boolean isFinished() {
        return Math.abs(delta) < 1 && Math.abs(drivetrain.getVelocities()[0]) < 0.015;
    }

    @Override
    public void end(boolean interrupted) {
        System.out.println("INFO: AlignShooter.end");
        drivetrain.stop();
    }
}

File: src/main/java/frc/robot/commands/Drivetrain_Commands/AlignShooter.kt
package frc.robot.commands.Drivetrain_Commands

import com.kauailabs.navx.frc.AHRS
import edu.wpi.first.math.MathUtil
import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import edu.wpi.first.wpilibj2.command.CommandBase
import edu.wpi.first.wpilibj2.command.CommandScheduler
import edu.wpi.first.wpilibj2.command.PIDCommand
import frc.robot.subsystems.DrivetrainFalcon
import frc.robot.util.PIDControllerDebug
import java.util.function.DoubleConsumer
import java.util.function.DoubleSupplier
import kotlin.math.abs
import kotlin.math.sign

class AlignShooter(private val imu: AHRS,
                   private val measurementSource: DoubleSupplier,
                   private val setpointSource: DoubleSupplier,
                   private val drivetrain: DrivetrainFalcon,
                   ) :
    CommandBase() {

    private var loopIdx = 0
    private var delta = 0.0

    override fun initialize() {
        println("INFO: AlignShooter is being called")
        delta = MathUtil.inputModulus(setpointSource.asDouble - measurementSource.asDouble, -180.0, 180.0)
    }

    override fun execute() {
        delta = MathUtil.inputModulus(setpointSource.asDouble - measurementSource.asDouble, -180.0, 180.0)


//        SmartDashboard.putNumber("Delta", delta)
        val output = if (abs(delta) > 45) {
            sign(delta) * 0.575
        } else if(abs(delta) > 30) {
            sign(delta) * 0.375
        } else if(abs(delta) > 1) {
            sign(delta) * 0.24
        } else {
            0.0
        }
        drivetrain.drive(0.0, -output)

//        // Debugging values
//        loopIdx++
//        if (loopIdx == 10) {
//            loopIdx = 0
//            println("INFO: Yaw: ${m_measurement.asDouble}, Heading: ${m_setpoint.asDouble}, PidOut: $pidOut, FF: $ff, TotalOut: ${pidOut+ff}")
//            println("INFO: Yaw: ${m_measurement.asDouble}, Heading: ${m_setpoint.asDouble}, Delta: $delta")
//            println("INFO: Yaw: ${m_measurement.asDouble}, Heading: ${m_setpoint.asDouble}, Output value: ${pidOut + ff}")
//
//        }
    }

    override fun isFinished(): Boolean {
        return if(imu.isConnected && !imu.isCalibrating) {
            abs(delta) < 1 && abs(drivetrain.getVelocities()[0]) < 0.015
        } else {
            println("ERROR: The IMU is disconnected or is calibrating")
            CommandScheduler.getInstance().cancelAll()
            true
        }

    }

    override fun end(interrupted: Boolean) {
        println("INFO: AlignShooter end")
        drivetrain.stop()
    }
}